File: multiprocessors/examples/06-openmp/example7.c
// =================================================================
//
// File: example7.c
// Author(s):
// 10/10/2022
// speedup = 141.46/34.89= 4.0544  
// A01705291 Ramses Aguila Flores
// A01209414 David Flores Diaz
// Description: This file contains the code to brute-force all
//				prime numbers less than MAXIMUM using OpenMP.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "utils.h"
#include <omp.h>

#define MAXIMUM 1000000 //1e6

// implement your code
	double prime(int n){
   		int lim;
   		if (n < 2)
      		  return 0;
   		lim = sqrt(n);
   		for (int i = 2 ; i <= lim ; i++){
      			if (n % i == 0) {
         	  	  return 0;
      			}
   		}
   		return 1;
	}

	double sum (int max){
   		double suma=0;
		#pragma omp parallel for shared(max) reduction(+:suma)
  	 	for (int i = 0 ; i <= max ; i++){
      		  if (prime(i)){
         		suma += i;
      		  }
   		}
  		return suma;
	}

	int main(int argc, char* argv[]) {
		int i, *a;
		double ms;
   		double res;
		a = (int *) malloc(sizeof(int) * (MAXIMUM + 1));
		printf("At first, neither is a prime. We will display to TOP_VALUE:\n");
		for (i = 2; i < TOP_VALUE; i++) {
	  	  if (a[i] == 0) {
	  	  printf("%i ", i);
	  	  }
		}

		printf("\n");
		printf("Starting...\n");
		ms = 0;
		for (i = 0; i < N; i++) {
			start_timer();
    		res = sum(MAXIMUM);// call the implemented function
			ms += stop_timer();
		}

   		printf("RES=%f \n",res);
		printf("Expanding the numbers that are prime to TOP_VALUE:\n");
		for (i = 2; i < TOP_VALUE; i++) {
	  	  if (a[i] == 1) {
	  	  printf("%i ", i);
	  	  }
		}
		printf("\n");
		printf("avg time = %.5lf ms\n", (ms / N));

		free(a);
		return 0;
	}

File: multiprocessors/examples/04-java-threads/Example7.java
// =================================================================
//
// File: Example7.java
// Author(s):
// Ramses Aguila Flores A01705291
// David Flores Diaz A01209414
// speedup = 312.4 / 183.7 = 1.7005 = 170.05 %
// Description: This file contains the code to brute-force all
//				prime numbers less than MAXIMUM. The time this
//				implementation takes will be used as the basis to
//				calculate the improvement obtained with parallel
//				technologies.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
import java.math.*;

public class Example7 extends Thread {
	private static final int SIZE = 1_000_000;
        private int array[], start, end;
        private double result;

        // place your code here

        public Example7(int array[], int start, int end) {
                this.array = array;
                this.start = start;
                this.end = end;
                this.result = 0;
        }

        public double getResult() {
                return result;
        }

	private boolean prime(int num){
		if(num <2){
			return false;
		}
		int lim=(int)Math.sqrt(num);
		for (int i=2;i<=lim;i++){
			if(num % i == 0){
			return false;
			}
		}
		return true;
	}

        public void run() {
                result = 0;
                for(int i = start ; i < end ; i++) {
                        if(prime(i)){
                                result+=i;
                        }
                }
        }


        public static void main(String args[]){
                long startTime, stopTime;
                int array[], block;
                Example7 threads[];
                double ms;
                double result = 0;

                array = new int[SIZE];

                Utils.fillArray(array);
                Utils.displayArray("array", array);

                block = SIZE / Utils.MAXTHREADS;
                threads = new Example7[Utils.MAXTHREADS];

                System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
                ms = 0;

                for (int j = 1; j <= Utils.N; j++) {
                        for (int i = 0; i < threads.length; i++) {
                                if (i != threads.length - 1) {
                                        threads[i] = new Example7(array, (i * block), ((i + 1) * block));
                                } else {
                                        threads[i] = new Example7(array, (i * block), SIZE);
                                }
                        }

                        startTime = System.currentTimeMillis();
                        for (int i = 0; i < threads.length; i++) {
                                threads[i].start();
                        }
                        /** -------- */
                        result = 0;
                        for (int i = 0; i < threads.length; i++) {
                                try {
                                        threads[i].join();
                                } catch (InterruptedException e) {
                                        e.printStackTrace();
                                }
                        }
                        stopTime = System.currentTimeMillis();
                        ms +=  (stopTime - startTime);

                }
                for (int i = 0; i < threads.length; i++) {
                result += threads[i].getResult();
                }

                System.out.printf("sum = %f\n", result);
                System.out.printf("avg time = %.5f ms\n", (ms / Utils.N));
        }
}

